packages/hardhat-core/src/internal/core/providers/http.ts

```diff
@@ -37,6 +39,9 @@ export class HttpProvider extends EventEmitter implements EIP1193Provider {
     private readonly _timeout = 20000
   ) {
     super();
+    const url = new URL(this._url);
+    this._path = url.pathname;
+    this._client = new Pool(url.toString().replace(url.pathname, ""));
```

Comment:
I'm not sure about this. Using the `username:password` with a node seems like something that could happen. OTOH, the undici docs [say](https://undici.nodejs.org/#/docs/api/Pool) "It should only include the protocol, hostname, and port"

@alcuadrado wdyt?

Reply:
This API was a bit surprising to me initially, but I guess it kinda makes sense for a connection pool. I would expect it to be created with just an origin, not an entire URL though.

I think we should still support that option, as it currently works, and dropping it is a breaking change.

This snippet shows how we can do it:

```js
const { Pool } = require("undici");
const { URL } = require("url");

async function main() {
  const url = new URL("http://admin:pass@127.0.0.1:8080");
  const pool = new Pool(url);

  const authheaderRequired = url.username !== "" && url.password !== "";

  let headers = {};
  if (authheaderRequired) {
    headers = {
      Authorization:
        "Basic " +
        new Buffer.from(url.username + ":" + url.password, "utf-8").toString(
          "base64"
        ),
    };
  }

  const req = await pool.request({
    path: "/",
    method: "GET",
    headers,
  });

  console.log(await req.body.text());
}

main().catch(console.error);
```

Reply:
Thanks for keeping this use case alive.  I took your suggestion in https://github.com/nomiclabs/hardhat/pull/2346/commits/8385d536fd64face5baf20cfc42a21bd4598cf27.

However, I took it with one caveat: we need to parse apart the URL string before passing it into the `Pool` constructor (regardless of whether we pass the `string` or a `URL` object), because if the URL includes a path, the `Pool` constructor will throw `InvalidArgumentError: invalid url`.

I have to admit that I haven't actually tested this :sweat_smile: either manually or in an automated way. (The existing test~s~ of `HttpProvider` ~are~ is quite thin: https://github.com/nomiclabs/hardhat/blob/master/packages/hardhat-core/test/internal/core/providers/http.ts )

Reply:
I tested the auth thing manually with my script and an express server